File: audio_ofdm_offline/sync_buffer.cpp
#include "sync_buffer.h"

SyncBuffer::SyncBuffer()
{
    _buffer = NULL;
    _numElem = 0;
    _alignIndex = 0;
}

SyncBuffer::~SyncBuffer()
{
    delete[] _buffer;
}

void SyncBuffer::setup(int numElem)
{
    delete[] _buffer;
    _numElem = numElem;
    _buffer = new int[3 * _numElem];
    reset();
}

void SyncBuffer::reset()
{
    _alignIndex = _numElem;
    for(int n = 0; n < 3 * _numElem; n++)
    {
       _buffer[n] = 0; 
    }
}

void SyncBuffer::insert(const int *data)
{
    for(int n = 0; n < _numElem; n++)
    {
        _buffer[n] = _buffer[n + _numElem];
        _buffer[n + _numElem] = data[n];
    }
}

void SyncBuffer::setAlignIndex(int index)
{
    _alignIndex = index;
}

const int* SyncBuffer::getAlignedBuffer()
{
  return _buffer + _alignIndex;
}

const int* SyncBuffer::getFullBuffer()
{
  return _buffer;
}

int SyncBuffer::getAlignIndex()
{
   return _alignIndex; 
}

int SyncBuffer::getNumElem()
{
   return _numElem; 
}

File: audio_ofdm_reference/SyncBuffer.m
classdef SyncBuffer
    properties
        buffer
        numElem
        alignIndex
    end
    methods
        function obj = SyncBuffer()
            obj.buffer     = 0;
            obj.numElem    = 0;
            obj.alignIndex = 0;
        end
        function obj = setup(obj, val)
            obj.numElem = val;
            obj.buffer = zeros(2 * val, 1);
            obj = reset(obj);
        end
        function obj = reset(obj)
            obj.alignIndex = obj.numElem + 1;
            obj.buffer(:) = 0;
        end
        function obj = insert(obj, buffer)
            obj.buffer(1:obj.numElem) = obj.buffer(obj.numElem + 1:end);
            obj.buffer(obj.numElem + 1:end) = buffer(:);
        end
        function obj = setAlignIndex(obj, val)
            obj.alignIndex = val;
        end
        function buffer = getAlignBuffer(obj, len)
            if (nargin < 2)
                len = obj.numElem;
            end
            buffer = obj.buffer(obj.alignIndex:obj.alignIndex + len - 1);
        end
        function buffer = getFullBuffer(obj, len)
            if (nargin < 2)
                len = length(obj.buffer);
            end
            buffer = obj.buffer(1:len);
        end
        function out = getAlignIndex(obj)
            out = obj.alignIndex;
        end
        function out = getNumElem(obj)
            out = obj.numElem;
        end
    end
end